### PR TITLE
Add unit tests to the updated Lambda code

### PR DESCRIPTION
**Pull request by creindle**
Addressing issue #4

### CODE REVIEW

**Comment by creindle** [comment]
When I run the newly created `cloudwatch.spec.js` file, I get an error saying `logger.createLogStream is not a function` for this specific [spot](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/src/cloudwatch.js#L50). When I dig deeper into the `cloudwatch.js` code, I see that the catch is being returned in the `getStreamInfo` function [here](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/src/cloudwatch.js#L31), so I'm a bit stuck at the `logger.describeLogStreams` [method](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/src/cloudwatch.js#L23). 

Should I add more information to the record being passed into the cloudwatch module? Or should I change the environment variable CLOUDWATCH_GROUP to pass this part?

**Comment by mfuroyama** [comment]
The solution here is to create a mock class for the CloudWatch logger interface of the AWS SDK.

Take a look at the [`s3-mock.js`](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/tests/s3-mock.js) module. You have to mock the interfaces provided by the object that is returned by the call to [`new AWS.CloudWatchLogs()`](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/src/cloudwatch.js#L14). So basically you'd need to create mock interfaces for `createLogStream`, `describeLogStreams` and `putLogEvents`.

Once you've created the `CloudwatchMock` class, you can use that in the `cloudwatch-spec.js` spec file, instead of the `S3Mock` class. That class was intended to be used to mock the calls made on an `S3` AWS object. Make sense?

Let me know if the instruction or the subject matter is unclear to you, or if you need help understanding the AWS or Jest mock mechanisms.

**Comment by creindle** [comment]
That makes MUCH more sense! Ok! I knew I had a big part missing hahahaha

**Comment by creindle** [comment]
Hey Mike, the issue that I have now is when I run the `cloudwatch.spec.js` test, the tests hang. I see that it's stuck on the promise from [logger.describeLogStreams](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/src/cloudwatch.js#L23) because when I print the mock contents, it contains the data from this method. 

I created the `CloudwatchLogsMock` class with the pattern laid in the S3 class. Did I create this correctly? This most likely would be the first step to seeing what the issue is.

**Comment by mfuroyama** [comment]
Yeah, it's probably hanging because the code is expecting the `describeLogStreams` to return a value. What the mock function is doing is simulating the functionality normally provided by the AWS SDK [CloudWatchLogs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudWatchLogs.html#describeLogStreams-property) interface.

So, if you look at the documentation, the call to `promise()` returns a Promise which will resolve with a JS object that contains an `uploadSequenceToken` if the log stream exists. You'll need to tweak the mock function to simulate this behavior. Right now, the Promise is resolving with nothing, which will cause the code to halt and probably crash. Make sense?

Sorry, I should have been more clear, maybe? When I said follow the pattern in the S3 mock, I didn't mean you should just cut-n-paste the mock class code. I intended for you look at the mock code as a guide, understand why the mock code was written in that way, understand how the mock interacts with the application code, then use that pattern and understanding to create a new mock and corresponding tests. 

Let me know if you need help understanding the concepts a little better. Also, let me know if you feel this issue is beyond your grasp, we can discuss breaking the issue up into smaller tasks.

**Comment by creindle** [comment]
In the S3 mock, why do you use this [line](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/tests/s3-mock.js#L21)?

**Comment by mfuroyama** [comment]
> In the S3 mock, why do you use this [line](https://github.com/vistadataproject/KinesisStreamTransform/blob/master/tests/s3-mock.js#L21)?

Because I want to be able to determine whether or not the Promise was actually being called. I can use `expect(mock.promise).toHaveBeenCalled()` to determine that

File: tests/cloudwatch-logs-mock.js [hunk 1]
```diff
@@ -21,13 +21,13 @@ class CloudwatchLogsMock {
...
+                    "logStreamName": data.logStreamNamePrefix,
                     "storedBytes": 0
                 },
                 {
                     "creationTime": 1433189873898,
                     "arn": "arn:aws:logs:us-west-2:0123456789012:log-group:my-logs:log-stream:20150601",
-                    "logStreamName": "20150601",
+                    "logStreamName": data.logStreamNamePrefix,
                     "storedBytes": 0
                 }
```
Comment (mfuroyama) [inline tests/cloudwatch-logs-mock.js]: If the `logStreamName` fields are the same, then the second object becomes redundant and unnecessary

**Comment by creindle** [comment]
Completed the tasks for this issue!

File: tests/cloudwatch-logs-mock.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    reset() {
+        this.data.splice(0);
+        this.promise.mockReset();
+    }
+
+    describeLogStreams(data) {
+        this.data.push(data);
+        const sampleResponse = {
+            "logStreams": [
```
Comment (mfuroyama) [inline tests/cloudwatch-logs-mock.js]: Has it been so long that we've forgotten to use single-quotes? 😆 

Also, when you build JS objects in Javascript code, you can forego the quotes on the field key names.

File: tests/cloudwatch-logs-mock.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        };
+    }
+
+    createLogStream(data) {
+        this.data.push(data);
+        const sampleResponse = `HTTP/1.1 200 OK
+x-amzn-RequestId: <RequestId>
+Content-Type: application/x-amz-json-1.1
+Content-Length: <PayloadSizeBytes>
+Date: <Date>`;
```
Comment (mfuroyama) [inline tests/cloudwatch-logs-mock.js]: This is okay to leave here, but really, when you think about it, there's really no need for this level of detail on the mock. For the purposes of the code we're writing, we don't depend on the contents of the `createLogStream`, just that it does something and it doesn't reject.

The teachable moment here is: when you're writing unit tests, always consider what it is, exactly, you're testing.

File: tests/cloudwatch-logs-mock.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    putLogEvents (data) {
+        this.data.push(data);
+        const sampleResponse = {
+            "nextSequenceToken": "test",
+            "rejectedLogEventsInfo": {
+                "expiredLogEventEndIndex": 1,
+                "tooNewLogEventStartIndex": 1,
+                "tooOldLogEventEndIndex": 1
```
Comment (mfuroyama) [inline tests/cloudwatch-logs-mock.js]: Double-quotes... you're breaking my ❤️ Christie... 😞

**Comment by mfuroyama** [comment]
What kind of coverage levels are you getting on these tests? Just wondering

File: tests/cloudwatch-logs-mock.js (hunk 2 above)
Comment (creindle) [reply tests/cloudwatch-logs-mock.js]: It's been sooo long...

File: tests/cloudwatch-logs-mock.js (hunk 3 above)
Comment (creindle) [reply tests/cloudwatch-logs-mock.js]: That's a good point!

**Comment by creindle** [comment]
```
---------------|----------|----------|----------|----------|-------------------|
File           |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
---------------|----------|----------|----------|----------|-------------------|
All files      |      100 |    96.15 |      100 |      100 |                   |
 cloudwatch.js |      100 |      100 |      100 |      100 |                   |
 index.js      |      100 |      100 |      100 |      100 |                   |
 s3.js         |      100 |      100 |      100 |      100 |                   |
 store.js      |      100 |       75 |      100 |      100 |                12 |
 stream.js     |      100 |      100 |      100 |      100 |                   |
---------------|----------|----------|----------|----------|-------------------|
```

This is the coverage now!

**Pull request merged**